### PR TITLE
COSE Signed API

### PR DESCRIPTION
**Pull request by nodh**
Adds a type parameter to `CoseSigned`. While not as thorough as for `JwsSigned`, it's still worth it to provide some form of guidance to clients as which payload to expect.

See also https://github.com/a-sit-plus/vck/pull/165

### CODE REVIEW

File: indispensable-cosef/src/commonMain/kotlin/at/asitplus/signum/indispensable/cosef/CoseSigned.kt [hunk 1]
```diff
@@ -1,83 +1,78 @@
...
     val signature: CryptoSignature by lazy {
         if (protectedHeader.value.usesEC() ?: unprotectedHeader?.usesEC() ?: (rawSignature.size < 2048))
             CryptoSignature.EC.fromRawBytes(rawSignature)
         else CryptoSignature.RSAorHMAC(rawSignature)
     }
 
-    fun serialize() = coseCompliantSerializer.encodeToByteArray(this)
+    fun serialize(): ByteArray = coseCompliantSerializer.encodeToByteArray(CoseSignedSerializer(), this)
+
+    fun getTypedPayload(deserializationStrategy: DeserializationStrategy<P>): KmmResult<P?> = catching {
```
Comment (JesusMcCloud) [inline indispensable-cosef/src/commonMain/kotlin/at/asitplus/signum/indispensable/cosef/CoseSigned.kt:53]: This should be lazy and a test that is not already based on a bytearray would be nice

**Review by JesusMcCloud — CHANGES_REQUESTED** [review]
Why does this behave differently from JwsSigned, that has the typed payload as property? should not the seriializer handle conversion from/to bytestringwrapper?

**Comment by JesusMcCloud** [comment]
> Why does this behave differently from JwsSigned, that has the typed payload as property? should not the seriializer handle conversion from/to bytestringwrapper?

In VCK, there's bytestringwrapper all over the place, i think we can do away with it and add this wrapping logic into the serializer, while we're at it

**Comment by JesusMcCloud** [comment]
@n0900 can this be made more convenient: as in: directly pass the typed payload to the constructor and have the serializer handle the byte string wrapping? I know it's different from your serialization foo, because it can by any type, but still…

**Comment by nodh** [comment]
Best I can do with my knowledge of kotlinx serialization ... `CoseSignedSerializer` can't handle the payload directly, because it doesn't know about the type of the payload, since it's referenced by annotation on `CoseSigned`. In contrast, `JwsSigned` itself is not annotated with `@Serializable`, so it is easier there. I'm happy for any suggestions on how to solve the shortcomings.

**Review by JesusMcCloud — APPROVED** [review]

**Pull request merged**